builder/common/step_create_ssm_tunnel.go

```diff
@@ -75,15 +84,70 @@ func (s *StepCreateSSMTunnel) Run(ctx context.Context, state multistep.StateBag)
 			LocalPort:  s.LocalPortNumber,
 			Region:     s.Region,
 		}.Start(ssmCtx, ui)
-
 		if err != nil {
 			ui.Error(fmt.Sprintf("ssm error: %s", err))
 		}
 	}()
 
+	var privateKeys [][]byte
+	if len(s.CommunicatorConfig.SSH.SSHPrivateKey) != 0 {
+		privateKeys = append(privateKeys, s.CommunicatorConfig.SSH.SSHPrivateKey)
+	}
+	if s.CommunicatorConfig.SSH.SSHPrivateKeyFile != "" {
+		privateKey, err := s.CommunicatorConfig.ReadSSHPrivateKeyFile()
+		if err != nil {
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		privateKeys = append(privateKeys, privateKey)
+	}
+
+	if len(privateKeys) > 0 {
+		ui.Say("Uploading SSH public key to instance")
+	}
+	for _, privateKey := range privateKeys {
+		err := s.sendUserSSHPublicKey(instance, privateKey)
+		if err != nil {
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+	}
+
 	return multistep.ActionContinue
 }
 
+func (s *StepCreateSSMTunnel) sendUserSSHPublicKey(
+	instance *ec2.Instance,
+	privateKey []byte,
+) error {
+	publicKey, err := sshkey.PublicKeyFromPrivate(privateKey)
+	if err != nil {
+		return fmt.Errorf("Error getting public key from private key: %s", err)
+	}
+	log.Printf("public key: %s", publicKey)
+	svc := ec2instanceconnect.New(s.AWSSession)
+	input := &ec2instanceconnect.SendSSHPublicKeyInput{
+		AvailabilityZone: aws.String(*instance.Placement.AvailabilityZone),
+		InstanceId:       aws.String(*instance.InstanceId),
+		InstanceOSUser:   aws.String(s.CommunicatorConfig.SSH.SSHUsername),
+		SSHPublicKey:     aws.String(strings.TrimSuffix(string(publicKey), "\n")),
+	}
+	log.Printf("Sending public key to instance: %s", input)
```

Comment:
This log will probably be verbose, I suggest we either cherry pick the instance ID (or anything that makes sense here) and only print this out